### PR TITLE
test(fee): fee market tests

### PR DESCRIPTION
**Pull request by andrussal**
## 1. What does this PR implement?

This PR adds fee e2e tests.

It also includes 4 unit tests which currently fail, because it seems that the current logic fails to adapt to market changes properly (prices don't rise under demand and don't recover after quiet periods — failure messages have the details). 

There is also an issue where it seems the storage price is never updated at all — usage doesn't reach the price update. The PR includes a proposed fix for it.

## 2. Does the code have enough context to be clearly understood?
Yes

## 3. Who are the specification authors and who is accountable for this PR?
@andrussal 

## 4. Is the specification accurate and complete?
N/A

## 5. Does the implementation introduce changes in the specification?
N/A

## Checklist

* [x] 1. The PR title follows the Conventional Commits [specification](https://www.notion.so/How-to-open-a-Pull-Request-215261aa09df80deb538ed59f5e4e0b5).
* [x] 2. Description added.
* [x] 3. Context and links to Specification document(s) added.
* [x] 4. Main contact(s) (developers and specification authors) added
* [x] 5. Implementation and Specification are 100% in sync including changes. This is critical.
* [x] 6. Link PR to a specific milestone.
* [x] 7. New or changed logs were reviewed against the [logging guidelines](https://github.com/logos-blockchain/logos-blockchain/blob/master/CONTRIBUTING.md#logging-guidelines).

### CODE REVIEW

File: tests/cucumber_tests/features/fees.feature [hunk 1]
```diff
@@ -40,3 +40,89 @@ Feature: Fees
     And I fund an inscription transaction on node "NODE_1" as "FUNDED_INSCRIPTION"
     Then transaction "FUNDED_INSCRIPTION" is included on node "NODE_1" in 120 seconds
     Then I stop all nodes
+
+  # The execution gas price after every block must equal the spec's calculation,
+  # recomputed here independently. Storage usage must also reach the pricing
+  # update and move its price at an epoch boundary. Together these guard the
+  # consensus math and that the storage market remains connected to network
+  # usage; price viability is covered by ledger market-liveness tests.
+  @transactions_ci
```
Comment (hansieodendaal) [inline tests/cucumber_tests/features/fees.feature]: I think we can add a ` @fees_ci` tag now and run them in their own cucumber ci task, although it has to do with transactions, it is a separate topic.

File: tests/cucumber_tests/features/fees.feature [hunk 2]
```diff
@@ -40,3 +40,89 @@ Feature: Fees
...
+      | node_name | account_index | wallet_name | connected_to |
+      | NODE_1    | 1             | WALLET_1A   |              |
+      | NODE_1    | 2             | WALLET_2A   |              |
+    When node "NODE_1" is at height 1 in 180 seconds
+    And I submit an exactly funded self-transfer from wallet "WALLET_1A" via node "NODE_1" as "TRAFFIC_A"
+    And I submit an exactly funded self-transfer from wallet "WALLET_2A" via node "NODE_1" as "TRAFFIC_B"
+    And I record per-block gas prices on node "NODE_1" for 125 slots in 300 seconds
+    Then transaction "TRAFFIC_A" is included on node "NODE_1" in 60 seconds
+    And transaction "TRAFFIC_B" is included on node "NODE_1" in 60 seconds
+    And recorded execution gas prices follow the execution market spec reference
```
Comment (hansieodendaal) [inline tests/cucumber_tests/features/fees.feature:73]: As a test designer, what ensures that the detailed comment above this test is asserted? I would have expected to see some assertion about "epoch boundary" before.

File: tests/cucumber_tests/features/fees.feature [hunk 3]
```diff
@@ -40,3 +40,89 @@ Feature: Fees
...
+      | 4             | 1           | 2            |
+    And I have a cluster with capacity of 1 nodes
+    And I start nodes with wallet resources:
+      | node_name | account_index | wallet_name | connected_to |
+      | NODE_1    | 1             | FUNDER_A    |              |
+      | NODE_1    | 2             | FUNDER_B    |              |
+      | NODE_1    | 3             | RECIPIENT_A |              |
+      | NODE_1    | 4             | RECIPIENT_B |              |
+    When node "NODE_1" is at height 2 in 240 seconds
+    And I concurrently fund transfers of 500 LGO from wallets "FUNDER_A" and "FUNDER_B" to wallets "RECIPIENT_A" and "RECIPIENT_B" via node "NODE_1" as "TX_A" and "TX_B"
```
Comment (hansieodendaal) [inline tests/cucumber_tests/features/fees.feature]: Question about this step
```
    And I concurrently fund transfers of 500 LGO from wallets "FUNDER_A" and "FUNDER_B" to wallets "RECIPIENT_A" and "RECIPIENT_B" via node "NODE_1" as "TX_A" and "TX_B"
```
See related comment at `fn concurrently_fund_transfers`
Maybe we should generalize this step to take arguments as a table, then we are not limited by size two.
An interesting take on this _for a different PR and scenario and test case_ would be to choose different nodes for each transfer, or to submit different transactions from the same wallet to recipient(s) using different nodes.

File: tests/cucumber_tests/features/fees.feature [hunk 4]
```diff
@@ -40,3 +40,89 @@ Feature: Fees
     And I fund an inscription transaction on node "NODE_1" as "FUNDED_INSCRIPTION"
     Then transaction "FUNDED_INSCRIPTION" is included on node "NODE_1" in 120 seconds
     Then I stop all nodes
+
+  # The execution gas price after every block must equal the spec's calculation,
```
Comment (hansieodendaal) [inline tests/cucumber_tests/features/fees.feature:44]: These are nice _fees_ test additions. I think we could enhance them by also using the `NODE_1` built-in wallet in the mix, to make sure we get what we expect.

File: tests/src/common/fee_spec.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
+//! Shared helpers for the fee e2e scenarios.
+//!
+//! `ExecutionMarketReference` and the `SPEC_*` constants are the spec's price
+//! update written out again by hand from the spec documents (not imported
+//! from node code), so the conformance scenario notices if the node's
+//! pricing math changes. The rest are plain helpers for building and
+//! checking fee-paying transactions.
```
Comment (hansieodendaal) [inline tests/src/common/fee_spec.rs:7]: Spec reference perhaps?

File: tests/cucumber_tests/features/fees.feature (hunk 4 above)
Comment (hansieodendaal) [inline tests/cucumber_tests/features/fees.feature:44]: I am missing a step in a test in this feature that purposefully add a tip that would be required for fee increase at epoch boundary? Only cases where we have zero tip and a shortfall tip, not a positive tip.

File: tests/src/cucumber/steps/fees/actions.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        }
+    }
+
+    Err("blocks stream ended before the requested slots were observed".to_owned())
+}
+
+/// The streamed block plus any ancestors the stream subscription missed,
+/// oldest first: walks parent links back until it reconnects with the last
+/// recorded block.
+async fn block_and_missed_ancestors(
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/fees/actions.rs:214]: We could also use `/cryptarchia/blocks_range"` endpoint from genesis to the current tip/slot.

File: tests/src/cucumber/steps/fees/actions.rs [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        slot: u64::from(block.header.slot),
+        execution_price: prices.execution_base_gas_price.into_inner(),
+        storage_price: prices.storage_gas_price.into_inner(),
+        block_execution_gas: fee_spec::spec_block_execution_gas(block)?,
+    })
+}
+
+/// Fires two wallet transfer requests at the same time, funded from the same
+/// key set, and records both transactions under their aliases.
+pub async fn concurrently_fund_transfers(
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/fees/actions.rs]: Maybe this could be a general vec, not limited by size two.
`concurrently_fund_n_transfers`

File: tests/src/cucumber/steps/fees/assertions.rs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    info!(
+        target: TARGET,
+        "Storage gas price responded to usage over {} blocks", records.len(),
+    );
+
+    Ok(())
+}
+
+/// Checks two included transactions spent no common ledger input.
+pub async fn transactions_spent_disjoint_inputs(
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/fees/assertions.rs]: Can this ever fail? Do consensus, mempool and block proposal rules not guarantee this?

File: tests/src/cucumber/steps/fees/assertions.rs [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+    for _ in 0..1_000 {
+        if Some(cursor) == genesis_id {
+            break;
+        }
+
+        let Some(block) = client
+            .block(&cursor)
+            .await
+            .map_err(|source| format!("block fetch failed: {source}"))?
+        else {
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/fees/assertions.rs]: We could use the `"/cryptarchia/blocks_range"` endpoint for this

File: tests/src/cucumber/steps/fees/assertions.rs [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+/// Per the [1.0.0] Storage Market spec, usage recorded during an epoch
+/// drives the price update at its boundary. Guards that the live storage
+/// market remains connected to network usage.
+pub fn storage_prices_respond_to_usage(world: &CucumberWorld, step: &Step) -> StepResult {
+    let records = recorded_gas_prices(world, step)?;
+
+    let first_price = records[0].storage_price;
+    if records
+        .iter()
+        .all(|record| record.storage_price == first_price)
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/fees/assertions.rs]: Maybe we could use a stronger assertion here? Otherwise, it could pass for either 1 → 2 or 1 → 0.
```suggestion
        .any(|record| record.storage_price > first_price)
```

File: tests/src/common/fee_spec.rs [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    // which depends on the fee - so recompute until the fee stops changing.
+    let mut fee = fee_for_output(utxo.note.value);
+    loop {
+        let recomputed = fee_for_output(output_for_fee(fee));
+
+        if recomputed == fee {
+            break;
+        }
+        fee = recomputed;
+    }
```
Comment (hansieodendaal) [inline tests/src/common/fee_spec.rs:189]: Maybe we can add a fail-safe exit condition here, guarding against an encoding-size boundary that could theoretically oscillate between two fee values and hang the scenario.

File: tests/src/common/fee_spec.rs [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    },
+};
+use lb_testing_framework::configs::wallet::WalletAccount;
+
+use crate::common::wallet::transfer_proofs_for_funded_wallet_tx;
+
+/// [1.0.0] Execution Market: base fee starts at 1.
+pub const SPEC_GENESIS_EXECUTION_PRICE: u128 = 1;
+/// [1.0.0] Storage Market: storage price starts at 1.
+pub const SPEC_GENESIS_STORAGE_PRICE: u64 = 1;
```
Comment (hansieodendaal) [inline tests/src/common/fee_spec.rs]: Is this still used?

File: tests/src/common/fee_spec.rs [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+/// The mandatory fee of a transaction at the given prices: execution gas *
+/// base price + encoded signed size * storage price ([1.5.0] Mantle).
+///
+/// # Panics
+///
+/// Panics on gas overflow, which small test transactions cannot hit.
+#[must_use]
+pub fn fee_of<State: VerificationState>(tx: &SignedMantleTx<State>, prices: &GasPrices) -> u64 {
```
Comment (hansieodendaal) [inline tests/src/common/fee_spec.rs]: Is this still used?

**Review by hansieodendaal — COMMENTED** [review]
Very nice addition, good job! 

I left some comments to consider/answer, and I think the PR description could also be out of date now that issues identified by this PR have been addressed and merged to master.

File: tests/cucumber_tests/features/fees.feature (hunk 1 above)
Comment (andrussal) [reply tests/cucumber_tests/features/fees.feature]: > I think we can add a ` @fees_ci` tag now and run them in their own cucumber ci task, although it has to do with transactions, it is a separate topic.

Good idea. I’ll move these scenarios to a dedicated @fees_ci suite so failures and artifacts are reported separately.

File: tests/cucumber_tests/features/fees.feature (hunk 2 above)
Comment (andrussal) [reply tests/cucumber_tests/features/fees.feature:73]: > As a test designer, what ensures that the detailed comment above this test is asserted? I would have expected to see some assertion about "epoch boundary" before.

Good point. The price records indirectly showed the boundary, but the scenario did not assert it explicitly. I’ll add a step that verifies the recording crossed an epoch boundary before asserting the storage-price change.

File: tests/cucumber_tests/features/fees.feature (hunk 3 above)
Comment (andrussal) [reply tests/cucumber_tests/features/fees.feature]: > Question about this step
> 
> ```
>     And I concurrently fund transfers of 500 LGO from wallets "FUNDER_A" and "FUNDER_B" to wallets "RECIPIENT_A" and "RECIPIENT_B" via node "NODE_1" as "TX_A" and "TX_B"
> ```
> 
> See related comment at `fn concurrently_fund_transfers` Maybe we should generalize this step to take arguments as a table, then we are not limited by size two. An interesting take on this _for a different PR and scenario and test case_ would be to choose different nodes for each transfer, or to submit different transactions from the same wallet to recipient(s) using different nodes.

Agreed. I changed this to a table-based step, including the node for each transfer, so it is no longer limited to two requests.

File: tests/cucumber_tests/features/fees.feature (hunk 4 above)
Comment (andrussal) [reply tests/cucumber_tests/features/fees.feature:44]: > These are nice _fees_ test additions. I think we could enhance them by also using the `NODE_1` built-in wallet in the mix, to make sure we get what we expect.

Good point. The payment case now uses the built-in node wallet, while the inscription case uses the user wallet, so both paths are covered.

File: tests/src/common/fee_spec.rs (hunk 5 above)
Comment (andrussal) [reply tests/src/common/fee_spec.rs:7]: > Spec reference perhaps?

Added links to the corresponding spec versions.

File: tests/src/cucumber/steps/fees/actions.rs (hunk 6 above)
Comment (andrussal) [reply tests/src/cucumber/steps/fees/actions.rs:214]: > We could also use `/cryptarchia/blocks_range"` endpoint from genesis to the current tip/slot.

Good suggestion. The recorder now uses blocks_range for existing history and the live feed for new blocks.

File: tests/src/cucumber/steps/fees/assertions.rs (hunk 8 above)
Comment (andrussal) [reply tests/src/cucumber/steps/fees/assertions.rs]: > Can this ever fail? Do consensus, mempool and block proposal rules not guarantee this?

Yes, once both transactions are included, consensus already guarantees their inputs are disjoint. The intended check was the wallet’s pending-note reservation: if it reused an input, one transaction would not be included. So asserting that both transactions land is sufficient, and I removed the redundant input check.

File: tests/src/common/fee_spec.rs (hunk 11 above)
Comment (andrussal) [reply tests/src/common/fee_spec.rs:189]: > Maybe we can add a fail-safe exit condition here, guarding against an encoding-size boundary that could theoretically oscillate between two fee values and hang the scenario.

Good catch. I added cycle detection and a bounded number of attempts.

**Review by hansieodendaal — APPROVED** [review]
LGTM

Potential follow-up:
- https://github.com/logos-blockchain/logos-blockchain/pull/3111#discussion_r3673226510
- Update PR description
- Prove that the storage-price rise occurs at the epoch boundary - a regression can incorrectly change the storage price halfway through an epoch.
